compiler/noirc_evaluator/src/ssa/opt/inlining.rs

```diff
@@ -160,130 +189,280 @@ fn called_functions_vec(func: &Function) -> Vec<FunctionId> {
     ssa: &Ssa,
     inline_no_predicates_functions: bool,
     aggressiveness: i64,
-) -> BTreeSet<FunctionId> {
-    let mut brillig_entry_points = BTreeSet::default();
-    let mut acir_entry_points = BTreeSet::default();
-
-    if matches!(ssa.main().runtime(), RuntimeType::Brillig(_)) {
-        brillig_entry_points.insert(ssa.main_id);
-    } else {
-        acir_entry_points.insert(ssa.main_id);
-    }
+) -> InlineInfos {
+    let mut inline_infos = InlineInfos::default();
+
+    inline_infos.insert(
+        ssa.main_id,
+        InlineInfo {
+            is_acir_entry_point: ssa.main().runtime().is_acir(),
+            is_brillig_entry_point: ssa.main().runtime().is_brillig(),
+            ..Default::default()
+        },
+    );
 
+    // Handle ACIR functions.
     for (func_id, function) in ssa.functions.iter() {
-        if matches!(function.runtime(), RuntimeType::Brillig(_)) {
+        if function.runtime().is_brillig() {
             continue;
         }
 
         // If we have not already finished the flattening pass, functions marked
         // to not have predicates should be preserved.
         let preserve_function = !inline_no_predicates_functions && function.is_no_predicates();
         if function.runtime().is_entry_point() || preserve_function {
-            acir_entry_points.insert(*func_id);
+            inline_infos.entry(*func_id).or_default().is_acir_entry_point = true;
         }
 
-        for called_function_id in called_functions(function) {
-            if matches!(ssa.functions[&called_function_id].runtime(), RuntimeType::Brillig(_)) {
-                brillig_entry_points.insert(called_function_id);
+        // Any Brillig function called from ACIR is an entry into the Brillig VM.
+        for called_func_id in called_functions(function) {
+            if ssa.functions[&called_func_id].runtime().is_brillig() {
+                inline_infos.entry(called_func_id).or_default().is_brillig_entry_point = true;
             }
         }
     }
 
-    let times_called = compute_times_called(ssa);
+    let callers = compute_callers(ssa);
+    let times_called = compute_times_called(&callers);
 
-    let brillig_functions_to_retain: BTreeSet<_> = compute_functions_to_retain(
+    mark_brillig_functions_to_retain(
         ssa,
-        &brillig_entry_points,
-        &times_called,
         inline_no_predicates_functions,
         aggressiveness,
+        &times_called,
+        &mut inline_infos,
     );
 
-    acir_entry_points
-        .into_iter()
-        .chain(brillig_entry_points)
-        .chain(brillig_functions_to_retain)
+    inline_infos
+}
+
+/// Compute the time each function is called from any other function.
+fn compute_times_called(
+    callers: &BTreeMap<FunctionId, BTreeMap<FunctionId, usize>>,
+) -> HashMap<FunctionId, usize> {
+    callers
+        .iter()
+        .map(|(callee, callers)| {
+            let total_calls = callers.iter().fold(0, |acc, (_caller, calls)| acc + calls);
```

Comment:
```suggestion
            let total_calls = callers.values().sum();
```
nit: a little more idiomatic